Bigquery col alias override

douenergy opened the pull request:
## Which issue does this PR close?
https://github.com/Canner/wren-engine/issues/1089

## Rationale for this change
Add an additional unparser rule for every dialect to allow dialects to override column aliases.

## What changes are included in this PR?
Rewrite BigQuery column aliases.

## Are these changes tested?
In datafusion/sql/tests/cases/plan_to_sql.rs `roundtrip_statement_with_dialect`.

douenergy commented:
<img width="1101" alt="image" src="https://github.com/user-attachments/assets/673685b9-b8b1-49a3-b12e-2a3ed45cf357" />

goldmedal commented on datafusion/sql/src/unparser/dialect.rs:
```diff
@@ -253,7 +262,43 @@ impl Dialect for DefaultDialect {
+            | '[' | '\\' | ']' | '^' | '`' | '{' | '}' | '~' => true,
+            _ => false,
+        }) {
+            Ok(Some(self.col_alias_generator.next("col")))
```
I'm thinking about if the way to rename the alias is enough 🤔 . Consider the following case:
```
select * from (select a as \"a*\", b as \"b@\" from (select 1 as a , 2 as b)) where \"a*\" = 1
```
We will get the result like:
```
SELECT * FROM (SELECT `a` AS `col_1`, `b` AS `col_2` FROM (SELECT 1 AS `a`, 2 AS `b`)) WHERE (`a*` = 1)
```
The identifier in the where clause is missed. However, it's hard to handle the complex nested case in the current way 🤔 

I guess maybe we can try some encoding way? Just encode the invalid identifier to another valid name. Maybe something like:
```rust
fn encode_column_name(name: &str) -> String {
    let mut encoded_name = String::new();
    for c in name.chars() {
        match c {
            '!' | '"' | '$' | '(' | ')' | '*' | ',' | '.' | '/' | ';' | '?' | '@'
            | '[' | '\\' | ']' | '^' | '`' | '{' | '}' | '~' => {
                encoded_name.push_str(&format!("_{}", c as u32));
            }
            _ => encoded_name.push(c),
        }
    }
    encoded_name
}

/// Stdout:
/// Original name: col!name$with*special,chars
/// Encoded name: col_33name_36with_42special_44chars
fn main() {
    let original_name = "col!name$with*special,chars";
    let encoded_name = encode_column_name(original_name);
    println!("Original name: {}", original_name);
    println!("Encoded name: {}", encoded_name);
}
```
We can avoid handling the name mapping issue. WDYT?

douenergy replied:
Fair point 👍 , and a great test case

douenergy commented:
@goldmedal Thank you for the review. I have addressed your comments. I agree that encoding the column name is a better idea. 

Also, I will add a test case that includes a special character column name in the where clause.

goldmedal commented on datafusion/sql/src/unparser/dialect.rs:
```diff
@@ -253,7 +264,52 @@ impl Dialect for DefaultDialect {
+
+    fn col_alias_overrides(
+        &self,
+        _unparser: &Unparser,
```
Maybe we can remove this argument. It's not used in the method.
```suggestion
```

goldmedal commented on datafusion/sql/src/unparser/expr.rs:
```diff
@@ -675,21 +675,29 @@ impl Unparser<'_> {
     fn col_to_sql(&self, col: &Column) -> Result<ast::Expr> {
+        // Replace the column name if the dialect has an override
+        let col_name = match self.dialect.col_alias_overrides(self, &col.name)? {
+            Some(encoded_name) => encoded_name,
```
I prefer to rename it to `rewritten_name` because the encoding way is only implemented for BigQueryDialect.
```suggestion
            Some(rewritten_name) => rewritten_name,
```

goldmedal reviewed (approved):
Thanks @douenergy. looks good to me 👍 
I leave some minor suggestions.

goldmedal commented:
Thanks @douenergy
